`Bun.SQL`:  `columns` field for `SQLResultArray`

alexandre-lavoie opened the pull request:
### What does this PR do?

Improves: #18866

Exports a `columns` field on `SQLResultArray` providing additional context on columns returned by the query. This is mostly just binding PostgreSQL's  `RowDescription` that was already previously implemented.

This is meant to behave similarly to [postgres.js](https://github.com/porsager/postgres) `columns`. The main difference is our `type` uses the [internal OIDs](https://github.com/postgres/postgres/blob/master/src/include/catalog/pg_type.dat) while postgresql.js uses [JDBC's OIDs](https://jdbc.postgresql.org/documentation/publicapi/constant-values.html). This would make ours more accurate but could impact DX that are used to the old "worst" values.

Here is the result of a test query:

```js
const sql = Bun.SQL(connectionString);

// Using safe
console.log((await sql`select 1 as x, 2 as longer_name, 3 as sp3c141_n4m3`));

// Using unsafe
console.log((await sql.unsafe("select 1 as x, 2 as longer_name, 3 as sp3c141_n4m3")));
```

```
[
  {
    x: 1,
    longer_name: 2,
    sp3c141_n4m3: 3,
  }, count: 1, command: "SELECT", columns: [
    {
      name: "x",
      table: 0,
      number: 0,
      type: 23,
    }, {
      name: "longer_name",
      table: 0,
      number: 0,
      type: 23,
    }, {
      name: "sp3c141_n4m3",
      table: 0,
      number: 0,
      type: 23,
    }
  ]
]
```

- [x] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

- [x] I wrote TypeScript tests and they pass locally (`bun-debug test test/js/sql/sql.test.ts`)
- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] JSValue used outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed

Jarred-Sumner reviewed (changes requested):
Thank you for this PR

Instead of creating the columns array dynamically before each time we run the callback, could we make it a lazy getter on the handle and avoid the work of creating the array when it usually never will be used? 

To do that, you could add another item to this array in the class generator:

https://github.com/oven-sh/bun/blob/727b0853762669d533b7294e35668a4697080a30/src/bun.js/api/postgres.classes.ts#L80

Then, a getter function declaration in the list where the other functions are in that file

Then, a getter implementation in postgres.zig, probably somewhat similar to `getQueries` 

https://github.com/oven-sh/bun/blob/727b0853762669d533b7294e35668a4697080a30/src/sql/postgres.zig#L1389-L1398

alexandre-lavoie commented:
Great suggestion!

I refactored to the lazier approach. Given `columns` was already in use, decided I might as well implement the `statement` object to be closer to `postgres.js`'s implementation.

I also updated  the types so they should be abit cleaner on output. It could potentially be less strict with the optional fields but I'm worried something weird might happen in the future due to `INSERT` and `UPDATE`.

Also noticed that there is a formatting issue. I'm not sure why but locally my auto `zig-format` does not match the repo's.

Here is my latest test/result to cover all the changes:

```js
const sql = new Bun.SQL(connectionString);

const res = await sql`select 1 as x, 2 as longer_name, 3 as sp3c141_n4m3`;

console.log("Query:", res.query);
console.log("Statement:", res.statement);
console.log("Columns:", res.columns);
console.log("Value[0]:", res[0]);
```

```
Query: PostgresQuery {   executed  }
Statement: {
  string: "select 1 as x, 2 as longer_name, 3 as sp3c141_n4m3",
  columns: [
    {
      name: "x",
      table: 0,
      number: 0,
      type: 23,
    }, {
      name: "longer_name",
      table: 0,
      number: 0,
      type: 23,
    }, {
      name: "sp3c141_n4m3",
      table: 0,
      number: 0,
      type: 23,
    }
  ],
}
Columns: [
  {
    name: "x",
    table: 0,
    number: 0,
    type: 23,
  }, {
    name: "longer_name",
    table: 0,
    number: 0,
    type: 23,
  }, {
    name: "sp3c141_n4m3",
    table: 0,
    number: 0,
    type: 23,
  }
]
Value[0]: {
  x: 1,
  longer_name: 2,
  sp3c141_n4m3: 3,
}
```

Jarred-Sumner commented on src/sql/postgres.zig:
```diff
@@ -620,6 +620,38 @@ pub const PostgresSQLQuery = struct {
+
+            const obj = JSValue.createEmptyObject(globalObject, 2);
+
+            obj.put(globalObject, JSC.ZigString.static("string"), bun.String.init(this.query).toJS(globalObject));
```
```suggestion
            obj.put(globalObject, JSC.ZigString.static("string"), bun.String.createUTF8ForJS(globalObject, this.query));
```

alexandre-lavoie replied:
Perfect, was not aware of this one!

Maybe we should add to the `String.init`'s documentation: "If you are creating a string for JS, you probably want to use createUTF8ForJS"?

alexandre-lavoie replied:
Actually in this case, I misread and `this.query` is already a `bun.String` so I assume `this.query.toJS(globalObject)` would be more optimal.

RainyPixel commented:
Is any help required?

robobun commented:
Thanks for this PR, and sorry it sat for so long without a decision.

Closing it as superseded. Bun's native code has since been rewritten in Rust, and the files this branch changes (`src/sql/postgres.zig`, `src/sql/postgres/postgres_protocol.zig`, `src/bun.js/api/postgres.classes.ts`) no longer exist on `main`, so it can no longer be rebased.

The feature itself is being added against the Rust sources in #30037: `result.columns` and `result.statement` (`{ string, columns }`), with the same `{ name, type, table, number }` shape and `pg_type` OIDs as here, plus MySQL support. That PR references this one as the earlier attempt and includes the cases tested here (`columns` on tagged template and `unsafe()` results, `statement.string` for plain and parameterized queries, `statement.columns`). #26809 stays open to track the feature until it lands.
